src/main/java/smile/identity/core/Signature.java

```diff
@@ -1,46 +1,110 @@
 package smile.identity.core;
 
 import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
 
 import javax.crypto.BadPaddingException;
 import javax.crypto.Cipher;
 import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.Mac;
 import javax.crypto.NoSuchPaddingException;
+import javax.crypto.spec.SecretKeySpec;
+
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
 import java.security.*;
 import java.security.spec.X509EncodedKeySpec;
+import java.text.SimpleDateFormat;
 import java.util.Base64;
 
 public class Signature {
+
+    private static final String SHA_256 = "SHA-256";
+    private static final String HMAC_SHA_256 = "HmacSHA256";
+    public static final String SEC_KEY = "sec_key";
+    public static final String TIME_STAMP_KEY = "timestamp";
+    public static final String DATE_TIME_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'";
+    public static final String SIGNATURE_KEY = "signature";
+	
     private Integer partnerId;
-    private String api_key;
+    private String apiKey;
 
-    public Signature(String partner_id, String api_key) {
-        int partnerId = Integer.parseInt(partner_id);
-        this.partnerId = partnerId;
-        this.api_key = api_key;
+    public Signature(String partnerId, String apiKey) {
+        this.partnerId = Integer.parseInt(partnerId);
+        this.apiKey = apiKey;
+    }
+    
+	public String generate_signature() throws Exception {
+    	return generate_signature(System.currentTimeMillis());
```

Comment:
just to clarify on types, `signature` uses a string timestamp

Reply:
Definitely. 
The statement `mac.update(new SimpleDateFormat(DATE_TIME_FORMAT).format(timestamp).getBytes(StandardCharsets.UTF_8));` in the signature calculation method `generate_signature(Long timestamp)` enforces that; it format the `Long` timestamp value into a standard UTF-8 date-time string.